src/client/datascience/jupyter/kernels/kernelConnectionWrapper.ts

```diff
@@ -44,84 +44,128 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
     public readonly unhandledMessage = new Signal<this, IMessage<MessageType>>(this);
     public readonly anyMessage = new Signal<this, Kernel.IAnyMessageArgs>(this);
     public get serverSettings() {
-        return this.kernelConnection.serverSettings;
+        return (this.kernelConnection || this._previousKernelConnection).serverSettings;
     }
     public readonly disposed = new Signal<this, void>(this);
+    /**
+     * Use `kernelConnection` to access the value as its not a constant (can change over time).
+     * E.g. when restarting kernels or the like.
+     */
+    private _kernelConnection!: Kernel.IKernelConnection;
+    private readonly _previousKernelConnection: Kernel.IKernelConnection;
+    // private _isRestarting?: boolean;
+    private get kernelConnection(): undefined | Kernel.IKernelConnection {
```

Comment:
The rename is good, but you renamed the wrong one. ```get kernelConnection()``` can return undefined. It should be called 'getPossibleKernelConnection'. The other one (which is now named getPossibleKernelConnection) would make more sense as just ```get kernelConnection()``` because it always returns a connection

Reply:
Aah yes,